ext/dom/xpath_callbacks.c

```diff
@@ -0,0 +1,504 @@
+{
+	if (!php_dom_xpath_is_callback_name_valid_and_throw(name, name_validation, false)) {
+		return FAILURE;
+	}
+
+	php_dom_xpath_callback_ns *namespace = php_dom_xpath_callbacks_ensure_ns(registry, ns);
+	zend_fcall_info_cache* allocated_fcc = emalloc(sizeof(zend_fcall_info));
+	zend_fcc_dup(allocated_fcc, fcc);
+
+	zval registered_value;
+	ZVAL_PTR(&registered_value, allocated_fcc);
+
+	zend_hash_update(&namespace->functions, name, &registered_value);
+	if (register_func) {
+		register_func(ctxt, ns, name);
+	}
+
+	namespace->mode = PHP_DOM_REG_FUNC_MODE_SET;
+
+	return SUCCESS;
+}
+
+static zval *php_dom_xpath_callback_fetch_args(xmlXPathParserContextPtr ctxt, uint32_t param_count, php_dom_xpath_nodeset_evaluation_mode evaluation_mode, dom_object *intern, php_dom_xpath_callbacks_proxy_factory proxy_factory)
+{
+	if (param_count == 0) {
+		return NULL;
+	}
+
+	zval *params = safe_emalloc(param_count, sizeof(zval), 0);
+
+	/* Reverse order to pop values off ctxt stack */
+	for (zval *param = params + param_count - 1; param >= params; param--) {
+		xmlXPathObjectPtr obj = valuePop(ctxt);
+		ZEND_ASSERT(obj != NULL);
+		switch (obj->type) {
+			case XPATH_STRING:
+				ZVAL_STRING(param, (char *)obj->stringval);
+				break;
+			case XPATH_BOOLEAN:
+				ZVAL_BOOL(param, obj->boolval);
+				break;
+			case XPATH_NUMBER:
+				ZVAL_DOUBLE(param, obj->floatval);
+				break;
+			case XPATH_NODESET:
+				if (evaluation_mode == PHP_DOM_XPATH_EVALUATE_NODESET_TO_STRING) {
+					char *str = (char *)xmlXPathCastToString(obj);
+					ZVAL_STRING(param, str);
+					xmlFree(str);
+				} else if (evaluation_mode == PHP_DOM_XPATH_EVALUATE_NODESET_TO_NODESET) {
+					if (obj->nodesetval && obj->nodesetval->nodeNr > 0) {
+						array_init_size(param, obj->nodesetval->nodeNr);
+						zend_hash_real_init_packed(Z_ARRVAL_P(param));
+						for (int j = 0; j < obj->nodesetval->nodeNr; j++) {
+							xmlNodePtr node = obj->nodesetval->nodeTab[j];
+							zval child;
+							if (UNEXPECTED(node->type == XML_NAMESPACE_DECL)) {
+								xmlNodePtr nsparent = node->_private;
+								xmlNsPtr original = (xmlNsPtr) node;
+
+								/* Make sure parent dom object exists, so we can take an extra reference. */
+								zval parent_zval; /* don't destroy me, my lifetime is transfered to the fake namespace decl */
+								php_dom_create_object(nsparent, &parent_zval, intern);
+								dom_object *parent_intern = Z_DOMOBJ_P(&parent_zval);
+
+								php_dom_create_fake_namespace_decl(nsparent, original, &child, parent_intern);
+							} else {
+								proxy_factory(node, &child, intern, ctxt);
+							}
+							zend_hash_next_index_insert_new(Z_ARRVAL_P(param), &child);
```

Comment:
I had a cursory glance here, but I'm trusting you know what you are doing here.

Reply:
This is pre-existing, this is also in stable branches.